File: pkg/engine/requests.go
package engine

import (
	"errors"
	"fmt"
	"path"
	"slices"
	"sort"

	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/elements"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/logger"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/stack"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles"
)

// internal worker request used by GameEngine.CloneGame()
type cloneGameResponse struct {
	BaseResponse
	Clones []*game.Game
}
type cloneGameRequest struct {
	GameID      int
	ReservedIDs []int
	LogDir      string
}

func (req *cloneGameRequest) gameID() int {
	return req.GameID
}

func (req *cloneGameRequest) requiresWrite() bool {
	return false
}

func (req *cloneGameRequest) execute(g *game.Game) Response {
	count := len(req.ReservedIDs)
	clones := make([]*game.Game, count)
	resp := &cloneGameResponse{BaseResponse: BaseResponse{gameID: req.GameID}}

	if req.LogDir == "" {
		for i := range req.ReservedIDs {
			clones[i] = g.DeepClone()
		}
		resp.Clones = clones
		return resp
	}

	for i, id := range req.ReservedIDs {
		logFile := path.Join(req.LogDir, fmt.Sprintf("%v.jsonl", id))
		logger, err := logger.NewFromFile(logFile)
		if err != nil {
			resp.err = err
			return resp
		}

		clone, err := g.DeepCloneWithLog(&logger)
		if err != nil {
			resp.err = err
			return resp
		}

		clones[i] = clone
	}

	resp.Clones = clones
	return resp
}

type PlayTurnResponse struct {
	BaseResponse
	Game game.SerializedGame
}
type PlayTurnRequest struct {
	GameID int
	Move   elements.PlacedTile
}

func (resp *PlayTurnResponse) canRemoveGame() bool {
	err := resp.Err()
	return err != nil && errors.Is(err, stack.ErrStackOutOfBounds)
}

func (resp *PlayTurnResponse) canRemoveChildGames() bool {
	return resp.Err() == nil
}

func (req *PlayTurnRequest) gameID() int {
	return req.GameID
}

func (req *PlayTurnRequest) requiresWrite() bool {
	return true
}

func (req *PlayTurnRequest) execute(game *game.Game) Response {
	err := game.PlayTurn(req.Move)
	resp := &PlayTurnResponse{
		BaseResponse: BaseResponse{
			gameID: req.gameID(),
			err:    err,
		},
	}
	if err != nil {
		return resp
	}

	resp.Game = game.Serialized()
	return resp
}

// State of the game the request is made for.
// Currently, this is internally represented as a slice of moves to make on the base game
// which can be resolved to a Game by passing it that game.
// Eventually this should probably move to some kind of handles to cached state.
type GameState struct {
	serializedGame game.SerializedGame
	simulatedMoves []elements.PlacedTile
}

func (state *GameState) Serialized() game.SerializedGame {
	return state.serializedGame
}

func (state *GameState) resolve(baseGame *game.Game) (*game.Game, error) {
	if state == nil {
		return baseGame, nil
	}
	game := baseGame.DeepClone()
	for _, move := range state.simulatedMoves {
		if err := game.SwapCurrentTile(elements.ToTile(move)); err != nil {
			return nil, err
		}
		if err := game.PlayTurn(move); err != nil {
			return nil, err
		}
	}
	return game, nil
}

func (state *GameState) with(
	serializedGame game.SerializedGame,
	move elements.PlacedTile,
) *GameState {
	var simulatedMoves []elements.PlacedTile
	if state == nil {
		simulatedMoves = []elements.PlacedTile{move}
	} else {
		simulatedMoves = make([]elements.PlacedTile, 0, len(state.simulatedMoves)+1)
		simulatedMoves = append(simulatedMoves, state.simulatedMoves...)
		simulatedMoves = append(simulatedMoves, move)
	}
	return &GameState{
		serializedGame: serializedGame,
		simulatedMoves: simulatedMoves,
	}
}

// represents a tile and its probability to be drawn from the deck
type TileProbability struct {
	Tile        tiles.Tile
	Probability float32
}

// tiles have probabilities represented as a 32-bit float
type GetRemainingTilesResponse struct {
	BaseResponse
	TileProbabilities []TileProbability
}
type GetRemainingTilesRequest struct {
	BaseGameID   int
	StateToCheck *GameState
}

func (req *GetRemainingTilesRequest) gameID() int {
	return req.BaseGameID
}

func (req *GetRemainingTilesRequest) requiresWrite() bool {
	return false
}

func (req *GetRemainingTilesRequest) execute(baseGame *game.Game) Response {
	resp := &GetRemainingTilesResponse{BaseResponse: BaseResponse{gameID: req.gameID()}}
	game, err := req.StateToCheck.resolve(baseGame)
	if err != nil {
		resp.err = err
		return resp
	}

	remaining := game.GetRemainingTiles()
	total := float32(len(remaining))
	probabilities := []TileProbability{}
	for _, tile := range remaining {
		n := len(probabilities)
		i := sort.Search(n, func(i int) bool {
			if len(probabilities[i].Tile.Features) != len(tile.Features) {
				return len(probabilities[i].Tile.Features) > len(tile.Features)
			}
			for j := range tile.Features {
				a := probabilities[i].Tile.Features[j]
				b := tile.Features[j]
				if a.FeatureType != b.FeatureType {
					return a.FeatureType > b.FeatureType
				}
				if a.ModifierType != b.ModifierType {
					return a.ModifierType > b.ModifierType
				}
				if a.Sides != b.Sides {
					return a.Sides > b.Sides
				}
			}
			return true
		})
		if i < n && probabilities[i].Tile.ExactEquals(tile) {
			probabilities[i].Probability++
		} else {
			probabilities = slices.Insert(
				probabilities,
				i,
				TileProbability{Tile: tile, Probability: 1.0},
			)
		}
	}
	for i := range probabilities {
		probabilities[i].Probability /= total
	}

	resp.TileProbabilities = probabilities

	return resp
}

type MoveWithState struct {
	Move  elements.PlacedTile
	State *GameState
}

type GetLegalMovesResponse struct {
	BaseResponse
	Moves []MoveWithState
}
type GetLegalMovesRequest struct {
	BaseGameID   int
	StateToCheck *GameState
	TileToPlace  tiles.Tile
}

func (req *GetLegalMovesRequest) gameID() int {
	return req.BaseGameID
}

func (req *GetLegalMovesRequest) requiresWrite() bool {
	return false
}

func (req *GetLegalMovesRequest) execute(baseGame *game.Game) Response {
	resp := &GetLegalMovesResponse{BaseResponse: BaseResponse{gameID: req.gameID()}}
	baseGame, err := req.StateToCheck.resolve(baseGame)
	if err != nil {
		resp.err = err
		return resp
	}

	placements := baseGame.GetTilePlacementsFor(req.TileToPlace)
	resp.Moves = []MoveWithState{}
	for _, placement := range placements {
		for _, move := range baseGame.GetLegalMovesFor(placement) {
			game := baseGame.DeepClone()
			if err := game.SwapCurrentTile(elements.ToTile(move)); err != nil {
				resp.err = err
				return resp
			}
			if err := game.PlayTurn(move); err != nil {
				resp.err = err
				return resp
			}
			moveState := MoveWithState{
				Move:  move,
				State: req.StateToCheck.with(game.Serialized(), move),
			}
			resp.Moves = append(resp.Moves, moveState)
		}
	}

	return resp
}

File: python_bindings/carcassonne_engine/requests.py
from ._bindings import engine as _go_engine  # type: ignore[attr-defined] # no stubs
from .models import GameState, PlacedTile, SerializedGame, Tile

__all__ = (
    "BaseResponse",
    "PlayTurnRequest",
    "PlayTurnResponse",
    "GetRemainingTilesRequest",
    "GetRemainingTilesResponse",
    "TileProbability",
    "GetLegalMovesRequest",
    "GetLegalMovesResponse",
    "MoveWithState",
)


class BaseResponse:
    """
    Base game engine response class.

    This is an abstract base class and should not be instantiated.

    Some of the objects returned by the `GameEngine` objects subclass this.
    """

    __slots__ = ("_go_obj", "game_id", "exception")

    def __init__(self, go_obj: _go_engine.BaseResponse) -> None:
        self._go_obj = go_obj
        self.game_id = go_obj.GameID()
        self.exception: Exception | None = None
        try:
            go_obj.Err()
        except RuntimeError as exc:
            self.exception = exc


class PlayTurnRequest:
    """
    Game engine request for placing specified tile on the game with specified ID.
    """

    __slots__ = ("_go_obj", "_game_id", "_move")

    def __init__(self, *, game_id: int, move: PlacedTile) -> None:
        self._go_obj = _go_engine.PlayTurnRequest(GameID=game_id, Move=move._unwrap())
        self._game_id = game_id
        self._move = move

    def _unwrap(self) -> _go_engine.PlayTurnRequest:
        return self._go_obj

    @property
    def game_id(self) -> int:
        return self._game_id

    @property
    def move(self) -> PlacedTile:
        return self._move


class PlayTurnResponse(BaseResponse):
    """
    Game engine response for `PlayTurnRequest` instances.

    This class is not meant to be instantiated by users directly
    and should be considered read-only.

    The instances of this class are provided by the `GameEngine` objects.
    """

    __slots__ = ("game", "final_scores")

    def __init__(self, go_obj: _go_engine.PlayTurnResponse) -> None:
        super().__init__(go_obj)
        self.game = SerializedGame(go_obj.Game) if not self.exception else None
        self.final_scores: dict[int, int] | None = None
        if go_obj.FinalScores:
            self.final_scores = {k: v for k, v in go_obj.FinalScores.items()}


class GetRemainingTilesRequest:
    """
    Game engine request for getting the remaining tiles
    in the game with specified ID and state.
    """

    __slots__ = ("_go_obj", "_base_game_id", "_state_to_check")

    def __init__(
            self, *, base_game_id: int, state_to_check: GameState | None = None
    ) -> None:
        if state_to_check is not None:
            self._go_obj = _go_engine.GetRemainingTilesRequest(
                BaseGameID=base_game_id,
                StateToCheck=state_to_check._unwrap(),
            )
        else:
            # gopy bindings don't consider None as Go's nil for pointers
            self._go_obj = _go_engine.GetRemainingTilesRequest(
                BaseGameID=base_game_id,
            )
        self._base_game_id = base_game_id
        self._state_to_check = state_to_check

    def _unwrap(self) -> _go_engine.GetRemainingTilesRequest:
        return self._go_obj

    @property
    def base_game_id(self) -> int:
        return self._base_game_id

    @property
    def state_to_check(self) -> GameState | None:
        return self._state_to_check


class GetRemainingTilesResponse(BaseResponse):
    """
    Game engine response for `GetRemainingTilesRequest` instances.

    This class is not meant to be instantiated by users directly
    and should be considered read-only.

    The instances of this class are provided by the `GameEngine` objects.
    """

    __slots__ = ("tile_probabilities",)

    def __init__(self, go_obj: _go_engine.GetRemainingTilesResponse) -> None:
        super().__init__(go_obj)
        self.tile_probabilities = (
            [
                TileProbability(go_probability)
                for go_probability in go_obj.TileProbabilities
            ]
            if not self.exception
            else None
        )


class TileProbability:
    """
    A tile and its probability to be drawn from the deck.

    This class is not meant to be instantiated by users directly
    and should be considered read-only.

    The instances of this class are provided by the `GameEngine` objects.
    """

    __slots__ = ("tile", "probability")

    def __init__(self, go_obj: _go_engine.TileProbability) -> None:
        self.tile = Tile(go_obj.Tile)
        self.probability = go_obj.Probability


class GetLegalMovesRequest:
    """
    Game engine request for getting the legal moves for the placeable tile
    in the game with specified ID and state.
    """

    __slots__ = ("_go_obj", "_base_game_id", "_state_to_check", "_tile_to_place")

    def __init__(
            self,
            *,
            base_game_id: int,
            state_to_check: GameState | None = None,
            tile_to_place: Tile,
    ) -> None:
        if state_to_check is not None:
            self._go_obj = _go_engine.GetLegalMovesRequest(
                BaseGameID=base_game_id,
                StateToCheck=state_to_check._unwrap(),
                TileToPlace=tile_to_place._unwrap(),
            )
        else:
            # gopy bindings don't consider None as Go's nil for pointers
            self._go_obj = _go_engine.GetLegalMovesRequest(
                BaseGameID=base_game_id,
                TileToPlace=tile_to_place._unwrap(),
            )
        self._base_game_id = base_game_id
        self._state_to_check = state_to_check
        self._tile_to_place = tile_to_place

    def _unwrap(self) -> _go_engine.GetLegalMovesRequest:
        return self._go_obj

    @property
    def base_game_id(self) -> int:
        return self._base_game_id

    @property
    def state_to_check(self) -> GameState | None:
        return self._state_to_check

    @property
    def tile_to_place(self) -> Tile:
        return self._tile_to_place


class GetLegalMovesResponse(BaseResponse):
    """
    Game engine response for `GetLegalMovesRequest` instances.

    This class is not meant to be instantiated by users directly
    and should be considered read-only.

    The instances of this class are provided by the `GameEngine` objects.
    """

    __slots__ = ("moves",)

    def __init__(self, go_obj: _go_engine.GetLegalMovesResponse) -> None:
        super().__init__(go_obj)
        self.moves = (
            [MoveWithState(go_move) for go_move in go_obj.Moves]
            if not self.exception
            else None
        )


class MoveWithState:
    """
    A legal move and the game state it would result in.

    This class is not meant to be instantiated by users directly
    and should be considered read-only.

    The instances of this class are provided by the `GameEngine` objects.
    """

    __slots__ = ("move", "state")

    def __init__(self, go_obj: _go_engine.MoveWithState) -> None:
        self.move = PlacedTile(go_obj.Move)
        self.state = GameState(go_obj.State)


class GetMidGameScoreRequest:
    """
   Game engine request for getting points as if the game just finished
   in the game with specified ID and state.
   """
    __slots__ = ("_go_obj", "_base_game_id", "_state_to_check")

    def __init__(
            self, *, base_game_id: int, state_to_check: GameState | None = None
    ) -> None:
        if state_to_check is not None:
            self._go_obj = _go_engine.GetMidGameScoreRequest(
                BaseGameID=base_game_id,
                StateToCheck=state_to_check._unwrap(),
            )
        else:
            # gopy bindings don't consider None as Go's nil for pointers
            self._go_obj = _go_engine.GetMidGameScoreRequest(
                BaseGameID=base_game_id,
            )
        self._base_game_id = base_game_id
        self._state_to_check = state_to_check

    def _unwrap(self) -> _go_engine.GetMidGameScoreRequest:
        return self._go_obj

    @property
    def base_game_id(self) -> int:
        return self._base_game_id

    @property
    def state_to_check(self) -> GameState | None:
        return self._state_to_check


class GetMidGameScoreResponse(BaseResponse):
    """
    Game engine response for `GetMidGameScoreRequest` instances.

    This class is not meant to be instantiated by users directly
    and should be considered read-only.

    The instances of this class are provided by the `GameEngine` objects.
    """

    __slots__ = ("player_scores",)

    def __init__(self, go_obj: _go_engine.GetMidGameScoreResponse) -> None:
        super().__init__(go_obj)
        self.player_scores = (
            {
                score[0]: score[1]
                for score in go_obj.Scores
            }
            if not self.exception
            else None
        )
